client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java

```diff
@@ -419,89 +434,108 @@ protected void checkBlockSendResult(Set<Long> blockIds) {
+    }
+
+    resendFailedBlocks(resendCandidates);
   }
 
-  private void reSendFailedBlockIds(Set<TrackingBlockStatus> failedBlockStatusSet) {
+  private void resendFailedBlocks(Set<TrackingBlockStatus> failedBlockStatusSet) {
     List<ShuffleBlockInfo> reAssignSeverBlockInfoList = Lists.newArrayList();
     List<ShuffleBlockInfo> failedBlockInfoList = Lists.newArrayList();
     Map<ShuffleServerInfo, List<TrackingBlockStatus>> faultyServerToPartitions =
         failedBlockStatusSet.stream().collect(Collectors.groupingBy(d -> d.getShuffleServerInfo()));
-    Map<String, ShuffleServerInfo> faultyServers = shuffleManager.getReassignedFaultyServers();
     faultyServerToPartitions.entrySet().stream()
         .forEach(
             t -> {
               Set<String> partitionIds =
                   t.getValue().stream()
                       .map(x -> String.valueOf(x.getShuffleBlockInfo().getPartitionId()))
                       .collect(Collectors.toSet());
-              ShuffleServerInfo dynamicShuffleServer = faultyServers.get(t.getKey().getId());
+              ShuffleServerInfo dynamicShuffleServer =
+                  failoverShuffleServers.get(t.getKey().getId());
               if (dynamicShuffleServer == null) {
+                // todo: merge multiple requests into one.
                 dynamicShuffleServer =
                     reAssignFaultyShuffleServer(partitionIds, t.getKey().getId());
-                faultyServers.put(t.getKey().getId(), dynamicShuffleServer);
+                failoverShuffleServers.put(t.getKey().getId(), dynamicShuffleServer);
               }
 
               ShuffleServerInfo finalDynamicShuffleServer = dynamicShuffleServer;
-              failedBlockStatusSet.forEach(
-                  trackingBlockStatus -> {
-                    ShuffleBlockInfo failedBlockInfo = trackingBlockStatus.getShuffleBlockInfo();
-                    failedBlockInfoList.add(failedBlockInfo);
-                    reAssignSeverBlockInfoList.add(
-                        new ShuffleBlockInfo(
-                            failedBlockInfo.getShuffleId(),
-                            failedBlockInfo.getPartitionId(),
-                            failedBlockInfo.getBlockId(),
-                            failedBlockInfo.getLength(),
-                            failedBlockInfo.getCrc(),
-                            failedBlockInfo.getData(),
-                            Lists.newArrayList(finalDynamicShuffleServer),
-                            failedBlockInfo.getUncompressLength(),
-                            failedBlockInfo.getFreeMemory(),
-                            taskAttemptId));
-                  });
+              for (TrackingBlockStatus blockStatus : failedBlockStatusSet) {
+                ShuffleBlockInfo failedBlockInfo = blockStatus.getShuffleBlockInfo();
+                failedBlockInfoList.add(failedBlockInfo);
+                ShuffleBlockInfo newBlock =
```

Comment:
Why do we use a new block instead of reusing old block?

Reply:
Sounds good. Let me optimize this tomorrow

Reply:
Emm, I reserve this logic becuase the old block will be used in the other places.

Reply:
It probably involve the issues if we use a new block. It's not a good practice. When I saw the this pattern, I will worry about the release of data and management of data. Maybe you use other code to guarantee this, but when seeing similar code, I will worry about the issues which I mentioned. So if it's possible, I prefer not using this pattern.

Reply:
Make sense.